### PR TITLE
[Task-03] IgorFromUA Components in React

### PR DESCRIPTION
**Pull request by IgorFromUA**
### Acceptance Criteria for 1
 

- [x] Mentor is added to pr
- [x]  AC copied to pr description
- [x]  Link to deployed version is added to pr
- [x]  The ProductsList component correctly receives product data as props.
- [x]  ProductCard component created and renders a single product.
- [x]  Header and Footer elements are present.
- [x]  ProductList displays right on mobile devices

### Acceptance Criteria for 1*
 

- [x] The Header component has clickable elements that toggle the display between the About and ProductsList components.
- [x]  State management is handled in App.tsx, with state determining which component to show.
- [x]  LocalStorage added to save items in card after page realod
### [link to deployed](https://igorfromua.github.io/fe-react-2024/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,17 +1,74 @@
...
+            setGeneralCart((previousState) => {
+                const nextState = [...previousState, id];
+                setLocalStorageCart(nextState);
+                return nextState;
+            });
+        }
+    };
+
+    const handlerSetActivePage = ({ currentTarget }: React.MouseEvent) => {
+        const page = currentTarget.getAttribute('name');
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx]: it's extra logic with attributes. You can do this in simple way. `currentTarget` can be replaced with string that will be from button click.

File: src/App.tsx [hunk 2]
```diff
@@ -1,17 +1,74 @@
...
+            return page;
+        });
+    };
+
     return (
         <>
             <div className={styles.app}>
                 <div className={styles.center}>
-                    <HeaderComponent />
+                    <HeaderComponent quantityInCart={generalCart.length} handlerSetActivePage={(event) => handlerSetActivePage(event)} />
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx]: `handlerSetActivePage={(event) => handlerSetActivePage(event)}` - it's same like `handlerSetActivePage={handlerSetActivePage}`

File: src/components/header/Header.component.tsx [hunk 3]
```diff
@@ -9,49 +11,59 @@ import userAdd from '../../assets/userAdd-icon.svg';
...
+                    <li>
+                        <img src={line} alt="line" />
+                    </li>
+                    <li>
+                        <img src={moon} alt="dark theme" />
+                    </li>
+                </ul>
+            </div>
+            <nav className={styles.nav}>
+                <button onClick={(event) => handlerSetActivePage(event)} name="about" data-name="about">
```
Comment (vadym-zinchenko-moc) [inline src/components/header/Header.component.tsx]: You can delete extra attributes i think. And write it  `handlerSetActivePage('about')` for about and ` handlerSetActivePage('product')` for products.

File: src/components/productCard/ProductCard.component.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    isProductInCart: boolean;
+    id: number;
+    title: string;
+    images: string[];
+    price: number;
+}
+
+const ProductCard: FC<ProductCardPropsType> = ({ onClickCart, isProductInCart, id, title, images, price }) => {
+    function handleClickCart() {
+        onClickCart(id);
```
Comment (vadym-zinchenko-moc) [inline src/components/productCard/ProductCard.component.tsx]: you can use `onClickCart` like props, it's extra method here `handleClickCart `

File: src/components/productsList/ProductsList.component.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import ProductCard from '../productCard/ProductCard.component';
+
+import styles from './productsList.module.css';
+
+interface Props {
+    onClickCart: Function;
+    generalCart: number[];
+}
+
+export const ProducstList: FC<Props> = ({ onClickCart, generalCart }) => (
```
Comment (vadym-zinchenko-moc) [inline src/components/productsList/ProductsList.component.tsx]: ProductList component must get `products` like props here. You need to import this in parent component

File: src/App.tsx [hunk 6]
```diff
@@ -1,17 +1,74 @@
...
+                return nextState;
+            });
+        }
+    };
+
+    const handlerSetActivePage = ({ currentTarget }: React.MouseEvent) => {
+        const page = currentTarget.getAttribute('name');
+        setActivePage(() => {
+            localStorage.setItem('activePage', page as string);
+            return page;
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx]: you can delete this callback from  `setActivePage` and set just page  and `localStorage.setItem('activePage', page as string);` write just inside `handlerSetActivePage`

File: src/components/productsList/ProductsList.component.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import type { FC } from 'react';
+
+import { products } from '@/moc/moc';
+
+import ProductCard from '../productCard/ProductCard.component';
+
+import styles from './productsList.module.css';
+
+interface Props {
```
Comment (vadym-zinchenko-moc) [inline src/components/productsList/ProductsList.component.tsx]: all `interfaces` must be in another folder
Check it please in each component of project

File: src/App.tsx [hunk 8]
```diff
@@ -1,17 +1,74 @@
...
+        });
+    };
+
     return (
         <>
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx:58]: seems like extra fragment

File: src/App.tsx [hunk 9]
```diff
@@ -1,17 +1,74 @@
...
+        if (localStorage.getItem('activePage')) setActivePage(localStorage.getItem('activePage'));
+    }, []);
+
+    function getLocalStorageCart(): Id[] {
+        if (!localStorage.getItem('generalCart')) return [];
+        const localStorageCart = localStorage.getItem('generalCart')?.split(' ').map(Number);
+        if (
+            Array.isArray(localStorageCart) &&
+            localStorageCart.every((item) => typeof item === 'number') &&
+            localStorageCart !== undefined &&
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx]: seems like extra condition. If `localStorageCart ` === undefined  `Array.isArray(localStorageCart)` will return false no?

File: src/App.tsx [hunk 10]
```diff
@@ -1,17 +1,74 @@
+import React, { useEffect, useState } from 'react';
+
 import { AboutComponent } from './components/about/About.component.tsx';
 import { FooterComponent } from './components/footer/Footer.component.tsx';
 import { HeaderComponent } from './components/header/Header.component.tsx';
+import { ProducstList } from './components/productsList/ProductsList.component.tsx';
 
 import styles from './App.module.css';
 
+type Page = string | null;
```
Comment (vadym-zinchenko-moc) [inline src/App.tsx]: Better to move it into another folder with interfaces

File: src/components/header/Header.component.tsx [hunk 11]
```diff
@@ -9,49 +11,59 @@ import userAdd from '../../assets/userAdd-icon.svg';
...
-                        <li>
-                            <img src={moon} alt="dark theme" />
-                        </li>
-                    </ul>
-                </div>
-                <nav className={styles.nav}>
-                    <a href="http">About</a>
-                    <a href="http">Product</a>
-                </nav>
+interface Property {
```
Comment (vadym-zinchenko-moc) [inline src/components/header/Header.component.tsx]: move it into folder with interfaces

**Review by vadym-zinchenko-moc — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 1 above)
Comment (IgorFromUA) [reply src/App.tsx]: Done.

File: src/App.tsx (hunk 2 above)
Comment (IgorFromUA) [reply src/App.tsx]: Done.

File: src/components/header/Header.component.tsx (hunk 3 above)
Comment (IgorFromUA) [reply src/components/header/Header.component.tsx]: Done.

File: src/components/productCard/ProductCard.component.tsx (hunk 4 above)
Comment (IgorFromUA) [reply src/components/productCard/ProductCard.component.tsx]: Done.

File: src/components/productsList/ProductsList.component.tsx (hunk 5 above)
Comment (IgorFromUA) [reply src/components/productsList/ProductsList.component.tsx]: Done.

File: src/components/productsList/ProductsList.component.tsx (hunk 7 above)
Comment (IgorFromUA) [reply src/components/productsList/ProductsList.component.tsx]: Done.

File: src/App.tsx (hunk 8 above)
Comment (IgorFromUA) [reply src/App.tsx:58]: Done.

File: src/App.tsx (hunk 9 above)
Comment (IgorFromUA) [reply src/App.tsx]: Done.

File: src/App.tsx (hunk 10 above)
Comment (IgorFromUA) [reply src/App.tsx]: Done.

File: src/components/header/Header.component.tsx (hunk 11 above)
Comment (IgorFromUA) [reply src/components/header/Header.component.tsx]: Done.

File: src/App.tsx (hunk 6 above)
Comment (IgorFromUA) [reply src/App.tsx]: Done.

File: src/components/header/Header.component.tsx (hunk 11 above)
Comment (IgorFromUA) [reply src/components/header/Header.component.tsx]: thanks for the review and great tips.

**Review by vadym-zinchenko-moc — APPROVED** [review]

**Pull request merged**